src/nodes/dir.rs

```diff
@@ -129,15 +136,60 @@ impl Node for Dir {
         Ok(state.attr)
     }
 
-    fn lookup(&self, _name: &OsStr) -> NodeResult<(Arc<Node>, fuse::FileAttr)> {
-        Err(KernelError::from_errno(libc::ENOENT))
+    fn lookup(&self, name: &OsStr, ids: &IdGenerator, cache: &Cache)
+        -> NodeResult<(Arc<Node>, fuse::FileAttr)> {
+        let mut state = self.state.lock().unwrap();
+
+        if let Some(node) = state.children.get(name) {
+            let refreshed_attr = node.getattr()?;
+            return Ok((node.clone(), refreshed_attr))
+        }
+
+        let (child, attr) = {
+            let path = match state.underlying_path.as_ref() {
+                Some(underlying_path) => underlying_path.join(name),
+                None => return Err(KernelError::from_errno(libc::ENOENT)),
+            };
+            let fs_attr = fs::symlink_metadata(&path)?;
+            let node = cache.get_or_create(ids, &path, &fs_attr, self.writable);
```

Comment:
Is the IdGenerator ever used separately from the Cache? Maybe it should be a part of the Cache.

Reply:
Yes. It's used to generate handle identifiers which are not related to the cache, and possibly also generation numbers later on.